OBSDOCS-3312: Logging Documentation Restructure Plan (#110864)
openshift/openshift-docs: modules/understanding-loki-sizing.adoc

== How log volume affects sizing

Log volume directly impacts the resources required for Loki components:

Ingestion rate (GB/day):: Drives ingester CPU and memory requirements. Higher log volume requires more ingesters to handle the write load.

Retention period:: Drives object storage capacity. Retention multiplied by daily log volume determines total storage needed (for example, 100 GB/day × 30 days = 3 TB storage).

Query time range:: Affects querier resource requirements. Queries across longer time ranges require more memory and CPU to process results.

[id="ruler-component-cpu_{context}"]
== Ruler component CPU requirements

The ruler component evaluates alerting rules and retention policies continuously. Enabling the ruler significantly increases CPU requirements:

* Processes alerting rules against incoming log streams
* Evaluates retention policies for log deletion
* Scales based on: number of tenants × number of rules × log data volume

For example, the `1x.medium` size requires:
* *Without ruler:* 34 vCPUs
* *With ruler:* 42 vCPUs (+8 vCPUs for ruler evaluation)

[id="determining-initial-deployment-size_{context}"]
== Determining initial deployment size

> Because log volume cannot be predicted before installation, use these baseline recommendations:

Reviewer: ```suggestion
Because log volume is difficult to predict before installation, use these baseline recommendations:
```
Not impossible, e.g. if you already have similar clusters to compare with, if you have very detailed knowedge of your workload etc. However people usually don't have these so "difficult"

Small clusters (≤10 nodes)::
Start with `1x.extra-small` size

Medium to large clusters (>10 nodes)::
Start with `1x.medium` size

Single-node OpenShift (SNO) or edge::
Start with `1x.pico` size (minimum viable deployment)

[IMPORTANT]
====

> You cannot accurately estimate log volume before installation. Factors that make pre-installation sizing impossible:

Reviewer: ```suggestion
It is difficult to accurately estimate log volume before installation. Even if you have a good basis for prediction (e.g. other similar clusters) you should monitor actual log rates after installation. Factors that make it difficult to predict sizing accurately include:
```

* A single pod can produce 100 GB/day of logs
* Log volume varies dramatically based on application verbosity, not pod count
* Infrastructure logs (journald) and audit logs are not produced by containers
* Cluster activity (pod creation/deletion) affects log volume independently of workload size

> 

Reviewer: Openshift audit logs are a special (bad) case here. Not enabled by default, but important to many customers so worth a mention - they're the #1 cause of logs being to big to store. Maybe link to https://docs.redhat.com/en/documentation/openshift_container_platform/4.20/html/security_and_compliance/audit-log-policy-config

Start with a baseline size and adjust after monitoring actual log production.
====

[id="calculating-sizing-after-installation_{context}"]
== Calculating sizing after installation

After the logging stack is running, calculate the correct size based on observed metrics:

.Procedure

. Wait at least 24 hours after installation to capture representative log volume patterns.

. Query the actual log ingestion rate using Prometheus:
+
[source,terminal]
----
sum by (component_name) (rate(vector_component_sent_bytes_total{component_type="loki", component_kind="sink"}[5m]))
----

. Convert the rate to GB/day and compare against the sizing table:
+
[source,terminal]
----
# Example: If query returns 5242880 bytes/second
# = 5242880 * 86400 / 1073741824 = ~421 GB/day
# → Requires 1x.small (500 GB/day capacity)
----

. Update the `LokiStack` CR with the new size if needed:
+
[source,yaml]
----
spec:
  size: 1x.small  # Updated from initial 1x.extra-small
----

. Monitor CPU and memory utilization to ensure the deployment is not resource-constrained.

> [NOTE]

Reviewer: This is good. FYI it is also possible to deploy just the LogMetricExporter, which measures the log traffic on disk even if the logging stack is not deployed. However this is complicated to explain and much less accurate than measuring real loki ingestion so probably no need. The exporter is mostly intended for debugging log loss problems - comparing disk activity with collection. Forget I said anything [DELETE]

Reviewer 2: Is not only used for debugging log loss. We use every day to get the top log producers to see if something is wrong. The metrics coming from the LogFileMetricExporter are part of the Logging / Collection Dashboard.

In fact, it helps also when the customer is "log Forwarding" and Loki is not installed, as it gives as visibility of the logs produced (something different to the logs read and sent by Vector)

[id="log-volume-affects-sizing_{context}"]
== How log volume affects sizing

Log volume directly impacts the resources required for Loki components:

Ingestion rate (GB/day):: Drives ingester CPU and memory requirements. Higher log volume requires more ingesters to handle the write load.

Retention period:: Drives object storage capacity. Retention multiplied by daily log volume determines total storage needed (for example, 100 GB/day × 30 days = 3 TB storage).

Query time range:: Affects querier resource requirements. Queries across longer time ranges require more memory and CPU to process results.

[id="ruler-component-cpu_{context}"]
== Ruler component CPU requirements

The ruler component evaluates alerting rules and retention policies continuously. Enabling the ruler significantly increases CPU requirements:

* Processes alerting rules against incoming log streams
* Evaluates retention policies for log deletion
* Scales based on: number of tenants × number of rules × log data volume

For example, the `1x.medium` size requires:
* *Without ruler:* 34 vCPUs
* *With ruler:* 42 vCPUs (+8 vCPUs for ruler evaluation)

[id="sizing-scenarios_{context}"]

> == Example sizing scenarios

Reviewer 2: All this is wrong. From the logging perspective, the question of "big" or "small" is not a question of number of pods or logs, it's a question of number of logs produced and this question has not relation with pods or nodes sizing or the active users (as 1 single user can do heavy queries exhausting it and 10 users doing very small queries). Then, remove the relation of number of users/nodes/pods with the Logging sizing as this leads continually to wrong places and big discussions 

The sizing can be only known after having installed logging, not before. Once installed, it can be calculated using the metrics shared in this revision before.

Small development cluster::
* 10-20 namespaces with moderate logging
* 7-day retention
* 50 GB/day log volume
* 1-5 active users querying logs
* *Recommended size:* `1x.pico` or `1x.extra-small`

Production cluster::
* 50-100 namespaces with active workloads
* 30-day retention for compliance
* 200-500 GB/day log volume
* 24/7 monitoring with dashboards
* Multiple alert rules
* *Recommended size:* `1x.small` or `1x.medium` (with ruler)

Enterprise multi-cluster::
* 500+ namespaces across multiple clusters
* 90-day retention for audit compliance
* 2+ TB/day log volume
* Centralized log aggregation
* Extensive alerting and compliance rules
* *Recommended size:* `1x.medium` (with ruler) or larger custom deployment

[id="estimating-log-volume_{context}"]

> == Estimating your log volume

Reviewer 2: All inside here should be removed. As it was explained in this document, the log activity does not depend on the number of pods, in fact, the infra (journald) and audit are not produced by containers. 

For estimating, we have the metric:
~~~
sum by (component_name) (rate(vector_component_sent_bytes_total{component_type="loki", component_kind="sink"}[5m]))
~~~
